packages/odyssey-react-mui/src/labs/DataComponents/TableSettings.tsx

```diff
@@ -0,0 +1,138 @@
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { Dispatch, SetStateAction, memo, useCallback, useMemo } from "react";
+import { Checkbox as MuiCheckbox } from "@mui/material";
+import { useTranslation } from "react-i18next";
+
+import { densityValues } from "./constants";
+import { ListIcon, ShowIcon } from "../../icons.generated";
+import { MenuButton } from "../../MenuButton";
+import { MenuItem } from "../../MenuItem";
+import { TableProps, TableState } from "./componentTypes";
+
+export type TableSettingsProps = {
+  setTableState: Dispatch<SetStateAction<TableState>>;
+  tableOptions: TableProps;
+  tableState: TableState;
+};
+
+const TableSettings = ({
+  setTableState,
+  tableOptions,
+  tableState,
+}: TableSettingsProps) => {
+  const { t } = useTranslation();
+
+  const { hasChangeableDensity, hasColumnVisibility, columns } = tableOptions;
+  const { rowDensity, columnVisibility } = tableState;
+
+  const changeRowDensity = useCallback(
+    (value: (typeof densityValues)[number]) =>
+      (_event: React.MouseEvent<HTMLLIElement>) => {
+        // This is necessary to avoid linter errors, while the _event is necessary to satisfy the onClick type
+        if (process.env.NODE_ENV === "development") console.debug(_event);
+
+        setTableState((prevState) => ({
+          ...prevState,
+          rowDensity: value,
+        }));
+      },
+    [setTableState],
+  );
+
+  const changeColumnVisibility = useCallback(
+    (columnId: string) => (_event: React.MouseEvent<HTMLLIElement>) => {
+      // This is necessary to avoid linter errors, while the _event is necessary to satisfy the onClick type
+      if (process.env.NODE_ENV === "development") console.debug(_event);
+
+      setTableState((prevState) => ({
+        ...prevState,
+        columnVisibility: {
+          ...prevState.columnVisibility,
+          [columnId]: prevState.columnVisibility
+            ? prevState.columnVisibility[columnId] === false
+            : false,
+        },
+      }));
+    },
+    [setTableState],
+  );
+
+  const isColumnVisibilityChecked = useMemo(() => {
+    return columns.reduce(
+      (acc, column) => {
+        const isChecked = columnVisibility
+          ? columnVisibility[column.accessorKey!] !== false
+          : true;
+        acc[column.accessorKey!] = isChecked;
+        return acc;
+      },
+      {} as Record<string, boolean>,
+    );
+  }, [columns, columnVisibility]);
```

Comment:
The way this is written, the `useMemo` return value can be an implicit return.

Also, we can rewrite this reduce to make it more clear what's going on:

```suggestion
  const isColumnVisibilityChecked = useMemo(() => (
    columns.reduce(
      (columns, column) => ({
        ...columns,
        [column.accessorKey!]: (
          columnVisibility
          || (
            !columnVisibility
            && columnVisibility[column.accessorKey!] !== false
          )
        ),
      }),
      {} as Record<string, boolean>,
    );
  ), [columns, columnVisibility]);
```

The last thing I'll point out is that we're creating an object called `isColumnVisibilityChecked`. That name is for booleans, not objects.

Why not make it a `Set` and call it something like `visibleColumns`? A set is a data structure that tells you if something is or is not in an array. It's exactly what you want.